sensor_msgs/include/sensor_msgs/point_cloud_msg_wrapper.hpp

```diff
@@ -0,0 +1,455 @@
+}  // namespace detail
+
+///
+/// @brief      This class implements a point cloud message wrapper.
+///
+///             It is designed to simplify working with point cloud messages. The idea is that it
+///             wraps a point cloud message reference allowing for simple access and modification.
+///             Upon creation, the wrapper checks if the fields of the message correspond to the
+///             fields generated from the point type provided by the user. If this check has passed,
+///             then the fields are identical (including the offsets of the members) and it is safe
+///             to reinterpret the point cloud message as an array of PointT points. Note that due
+///             to these checks it is relatively expensive (the complexity of the checks is around
+///             O(n^2) where n is the number of field generators) to create this wrapper, thus it
+///             should be avoided in very tight scopes. This operation is not slow, but is
+///             significantly slower than a single element access. It is better to create the
+///             wrapper once per message in a function scope and batch all the read/write operations
+///             afterwards.
+///
+///             @warning This class wraps a raw reference, so the user is responsible to use it
+///             in such a way that the underlying point cloud message is not deleted before the
+///             wrapper.
+///
+///             For convenience, there are two typedefs: PointCloud2View and PointCloud2Modifier.
+///             - PointCloudView<PointT>{msg} wraps a constant point cloud and allows read-only
+///               access. Modification through this view is impossible.
+///             - PointCloud2Modifier<PointT>{msg} wraps a mutable message and allows read-write
+///               access to the underlying data.
+///             - PointCloud2Modifier<PointT>{msg, new_frame_id} initializes an empty mutable
+///               message. This constructor is to be used solely to initialize a new message and
+///               will throw if the point cloud message is already initialized (i.e., has non-zero
+///               number of fields).
+///
+/// @tparam     PointT           Type of point to use for message reading/writing.
+/// @tparam     CloudMsgT        Type of point cloud message.
+/// @tparam     FieldGenerators  A tuple of all field generators that allow generating Field structs
+///                              from members of a PointT struct. See
+///                              LIDAR_UTILS__DEFINE_FIELD_GENERATOR_FOR_MEMBER for more details.
+///                              The class provides a sane default value here, but a custom tuple
+///                              can be generated by the user by using
+///                              LIDAR_UTILS__DEFINE_FIELD_GENERATOR_FOR_MEMBER if needed.
+///
+template<typename PointT, typename CloudMsgT, typename FieldGenerators, std::size_t UPPER_BOUND>
```

Comment:
What is the reason behind having `CloudMsgT` as a template?

Reply:
Mostly to avoid code duplication for `const` and non-`const` versions of the point cloud type.